velox/dwio/parquet/tests/reader/ParquetTableScanTest.cpp

```diff
@@ -0,0 +1,142 @@
+    options.parseDecimalAsDouble = false;
+
+    auto plan = PlanBuilder()
+                    .setParseOptions(options)
+                    .tableScan(rowType, subfieldFilters, remainingFilter)
+                    .planNode();
+
+    assertQuery(plan, splits_, sql);
+  }
+
+  void
+  loadData(const std::string& filePath, RowTypePtr rowType, RowVectorPtr data) {
+    splits_ = {makeSplit(filePath)};
+    rowType_ = rowType;
+    createDuckDbTable({data});
+  }
+
+  std::string getExampleFilePath(const std::string& fileName) {
+    return facebook::velox::test::getDataFilePath(
+        "velox/dwio/parquet/tests/reader", "../examples/" + fileName);
+  }
+
+  std::shared_ptr<connector::hive::HiveConnectorSplit> makeSplit(
+      const std::string& filePath) {
+    return makeHiveConnectorSplits(
+        filePath, 1, dwio::common::FileFormat::PARQUET)[0];
+  }
+
+ private:
+  RowTypePtr getRowType(std::vector<std::string>&& outputColumnNames) const {
+    std::vector<TypePtr> types;
+    for (auto colName : outputColumnNames) {
+      types.push_back(rowType_->findChild(colName));
+    }
+
+    return ROW(std::move(outputColumnNames), std::move(types));
+  }
+
+  RowTypePtr rowType_;
+  std::vector<std::shared_ptr<connector::ConnectorSplit>> splits_;
+};
+
+TEST_F(ParquetTableScanTest, decimalSubfieldFilter) {
+  // decimal.parquet holds two columns (a: DECIMAL(5, 2), b: DECIMAL(20, 5)) and
+  // 20 rows (10 rows per group). Data is in plain uncompressed format:
+  //   a: [100.01 .. 100.20]
+  //   b: [100000000000000.00001 .. 100000000000000.00020]
+  std::vector<int64_t> unscaledShortValues(20);
+  std::iota(unscaledShortValues.begin(), unscaledShortValues.end(), 10001);
+  loadData(
+      getExampleFilePath("decimal.parquet"),
+      ROW({"a"}, {DECIMAL(5, 2)}),
+      makeRowVector(
+          {"a"},
+          {
+              makeShortDecimalFlatVector(unscaledShortValues, DECIMAL(5, 2)),
+          }));
+
+  assertSelectWithFilter(
+      {"a"}, {"a < 100.07"}, "", "SELECT a FROM tmp WHERE a < 100.07");
+  assertSelectWithFilter(
+      {"a"}, {"a <= 100.07"}, "", "SELECT a FROM tmp WHERE a <= 100.07");
+  assertSelectWithFilter(
+      {"a"}, {"a > 100.07"}, "", "SELECT a FROM tmp WHERE a > 100.07");
+  assertSelectWithFilter(
+      {"a"}, {"a >= 100.07"}, "", "SELECT a FROM tmp WHERE a >= 100.07");
+  assertSelectWithFilter(
+      {"a"}, {"a = 100.07"}, "", "SELECT a FROM tmp WHERE a = 100.07");
+  assertSelectWithFilter(
+      {"a"},
+      {"a BETWEEN 100.07 AND 100.12"},
+      "",
+      "SELECT a FROM tmp WHERE a BETWEEN 100.07 AND 100.12");
+
+  VELOX_ASSERT_THROW(
+      assertSelectWithFilter(
+          {"a"}, {"a < 1000.7"}, "", "SELECT a FROM tmp WHERE a < 1000.7"),
+      "Scalar function signature is not supported: lt");
+  VELOX_ASSERT_THROW(
+      assertSelectWithFilter(
+          {"a"}, {"a = 1000.7"}, "", "SELECT a FROM tmp WHERE a = 1000.7"),
+      "Scalar function signature is not supported: eq");
+}
```

Comment:
Can we add a test to check we throw exception if the precision/scale in filter does not match the one in file?

Reply:
I think that is good then